syntax = "proto3";

option optimize_for = LITE_RUNTIME;

package nil.xit.proto;

enum MessageType {
    MessageType_FrameRequest = 0;
    MessageType_FrameResponse = 1;
    MessageType_FrameCache = 2;
    MessageType_FrameLoaded = 3;
    MessageType_FrameSubscribe = 4;
    MessageType_FrameUnsubscribe = 5;
    MessageType_ValueRequest = 6;
    MessageType_ValueResponse = 7;
    MessageType_ValueUpdate = 8;
    MessageType_SignalRequest = 9;
    MessageType_SignalResponse = 10;
    MessageType_SignalNotify = 11;
    MessageType_FileRequest = 12;
    MessageType_FileResponse = 13;
}

message FrameRequest
{
    string id = 1;
}

message FrameResponse
{
    string id = 1;
    oneof value {
        string file = 2;
        string content = 3;
    }
}

message FrameCache {
    string id = 1;
    string content = 2;
    repeated FileInfo files = 3;
}

message FrameLoaded {
    string id = 1;
    optional string tag = 2;
}

message FrameSubscribe {
    string id = 1;
    optional string tag = 2;
}

message FrameUnsubscribe {
    string id = 1;
    optional string tag = 2;
}

message Value
{
    string id = 1;
    oneof value {
        bool value_boolean = 2;
        int64 value_number = 3;
        double value_double = 4;
        string value_string = 5;
        bytes value_buffer = 6;
    }
}

message ValueRequest
{
    string id = 1;
    optional string tag = 2;
}

message ValueResponse
{
    string id = 1;
    optional string tag = 2;
    repeated Value values = 3;
}

message ValueUpdate
{
    string id = 1;
    optional string tag = 2;
    Value value = 3;
}

message SignalRequest
{
    string id = 1;
}

message Signal
{
    string id = 1;
    optional string type = 2;
}

message SignalResponse
{
    string id = 1;
    optional string tag = 2;
    repeated Signal signals = 3;
}

message SignalNotify
{
    string frame_id = 1;
    optional string tag = 2;
    string signal_id = 3;
    oneof arg {
        bool arg_boolean = 4;
        int64 arg_number = 5;
        double arg_double = 6;
        string arg_string = 7;
        bytes arg_buffer = 8;
    };
}

message FileRequest
{
    string target = 1;
}

message FileResponse {
    string target = 1;
    string content = 2;
    bytes metadata = 3;
}

message FileInfo {
    string target = 1;
    bytes metadata = 2;
}